### PR TITLE
RATIS-2186. Raft log should not purge index lower than the log start index

### PR DESCRIPTION
**Pull request by ivandika3**
## What changes were proposed in this pull request?

Please refer to https://issues.apache.org/jira/browse/RATIS-2186

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/RATIS-2186

## How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by ivandika3** [comment]
cc: @szetszwo @SzyWilliam

**Review by szetszwo — COMMENTED** [review]
@ivandika3 , thanks a lot for working on this!

- Let's also change `SegmentedRaftLogCache` to handle the -1 case, which is the same as the `toDelete` list being empty.

```diff
+++ b/ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/SegmentedRaftLogCache.java
@@ -364,6 +364,9 @@ public class SegmentedRaftLogCache {
           list.addAll(segments);
           segments.clear();
           sizeInBytes = 0;
+        } else if (segmentIndex == -1) {
+          // nothing to purge
+          return null;
         } else if (segmentIndex >= 0) {
           // we start to purge the closedSegments which do not overlap with index.
           LogSegment overlappedSegment = segments.get(segmentIndex);
```

- See also the comments inlined and https://issues.apache.org/jira/secure/attachment/13072740/1175_review.patch

File: ratis-server/src/main/java/org/apache/ratis/server/raftlog/RaftLogBase.java [hunk 1]
```diff
@@ -326,6 +326,13 @@ public final CompletableFuture<Long> purge(long suggestedIndex) {
     if (suggestedIndex - lastPurge < purgeGap) {
       return CompletableFuture.completedFuture(lastPurge);
     }
+    long startIndex = getStartIndex();
+    if (suggestedIndex < startIndex) {
+      LOG.info("{}: purge is skipped since the suggested index {} is lower than " +
+              "log start index {}",
+          getName(), suggestedIndex, startIndex);
+      return CompletableFuture.completedFuture(lastPurge);
+    }
```
[the remark this answers is not in the document]
Comment (ivandika3) [reply ratis-server/src/main/java/org/apache/ratis/server/raftlog/RaftLogBase.java]: Thanks for the review. A lot clearer now. Updated as per the suggestion.

File: ratis-test/src/test/java/org/apache/ratis/server/raftlog/segmented/TestSegmentedRaftLog.java [hunk 2]
```diff
@@ -150,7 +150,7 @@ static SegmentedRaftLog newSegmentedRaftLog(RaftStorage storage, RaftProperties
         .build();
   }
 
-  private SegmentedRaftLog newSegmentedRaftLogWithSnapshotIndex(RaftStorage storage, RaftProperties properties,
+  SegmentedRaftLog newSegmentedRaftLogWithSnapshotIndex(RaftStorage storage, RaftProperties properties,
```
[the remark this answers is not in the document]
Comment (ivandika3) [reply ratis-test/src/test/java/org/apache/ratis/server/raftlog/segmented/TestSegmentedRaftLog.java]: Sure. Updated.

File: ratis-test/src/test/java/org/apache/ratis/server/raftlog/segmented/TestSegmentedRaftLog.java [hunk 3]
```diff
@@ -599,18 +600,34 @@ public void testPurgeOnClosedSegmentsWithPurgeGap() throws Exception {
...
+    // which causes suggested index to be lower than the start index
+    long purgePreservation = (segmentSize * (endTerm - startTerm )) + 100;
+    // if the suggested index is lower than the start index due to the purge preservation, we should not purge anything
+    purgeAndVerify(startTerm, endTerm, segmentSize, 1, endIndex, startIndex, startIndex, purgePreservation);
   }
 
   private void purgeAndVerify(int startTerm, int endTerm, int segmentSize, int purgeGap, long purgeIndex,
-      long expectedIndex) throws Exception {
-    List<SegmentRange> ranges = prepareRanges(startTerm, endTerm, segmentSize, 0);
+      long expectedIndex, long startIndex, long purgePreservation) throws Exception {
```
[the remark this answers is not in the document]
Comment (ivandika3) [reply ratis-test/src/test/java/org/apache/ratis/server/raftlog/segmented/TestSegmentedRaftLog.java]: Sure. Updated.

**Comment by ivandika3** [comment]
@szetszwo Thanks for the review and the suggestions. I have updated them accordingly.

> Let's also change SegmentedRaftLogCache to handle the -1 case, which is the same as the toDelete list being empty.

This is a good idea to prevent similar issue when calling `SgmentedRaftLogCache#purge`

File: ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/SegmentedRaftLogCache.java [hunk 4]
```diff
@@ -364,6 +364,9 @@ TruncationSegments purge(long index) {
           list.addAll(segments);
           segments.clear();
           sizeInBytes = 0;
+        } else if (segmentIndex == -1) {
```
Comment (OneSizeFitsQuorum) [inline ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/SegmentedRaftLogCache.java:367]: When segments is Empty, it seems code will go to `if (segmentIndex == -segments.size() - 1) {` ?

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Comment by szetszwo** [comment]
@ivandika3 , thanks a lot for working on this!

@OneSizeFitsQuorum , thanks for reviewing this!

**Comment by ivandika3** [comment]
@szetszwo @OneSizeFitsQuorum Thanks for the reviews and suggestions.

**Pull request merged**